src/service.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The agent pebble service module."""
+
+import logging
+import os
+import pwd
+import time
+from pathlib import Path
+
+from charms.operator_libs_linux.v0 import apt
+from charms.operator_libs_linux.v1 import systemd
+from jinja2 import Template
+
+from charm_state import State
+
+logger = logging.getLogger(__name__)
+AGENT_SERVICE_NAME = "jenkins-agent"
+AGENT_PACKAGE_NAME = "jenkins-agent"
+SYSTEMD_SERVICE_CONF_DIR = "/etc/systemd/system/jenkins-agent.service.d/"
+PPA_GPG_KEY_ID = "67393A94A577DC24"
+READINESS_CHECK_DELAY = 30
+
+
+class PackageInstallError(Exception):
+    """Exception raised when package installation fails."""
+
+
+class ServiceRestartError(Exception):
+    """Exception raised when failing to start the agent service."""
+
+
+class JenkinsAgentService:
+    """Jenkins agent service class.
+
+    Attrs:
+       is_active: Indicate if the agent service is active and running.
+    """
+
+    def __init__(self, state: State):
+        """Initialize the jenkins agent service.
+
+        Args:
+            state: The Jenkins agent state.
+        """
+        self.state = state
+
+    def _render_file(self, path: str, content: str, mode: int) -> None:
```

Comment:
Should this include a bit more error handling, e.g., if file interactions fail etc?

Reply:
I looked into it in details but I can't find any meaningful cases where we would want to have custom error handling for the `write_text` and `os.chmod` methods. I think any exception raised here would be picked up by ops and the charm will be set to error state. Can you maybe elaborate on what scenario you have in mind?

Reply:
Letting exceptions bubble up can be surprising to the caller. How are they supposed to know that a given function deep down could raise some sort of OS file error? Usually, exceptions that could occur within a function should be re-mapped to different exceptions and documented in the docstring so that the caller can decide what to do in case they occur. There are practicalities to consider, e.g., if there are some extremely rare exceptions or things like KeyboardInterrupt, but that guidance is a good starting point

Reply:
Got it, that makes sense and it's better to be consistent with our error handling. I'll add a blanket catch for exceptions in this method for now and remap them to custom ones